cmd/snap-bootstrap/triggerwatch/evdev.go

```diff
@@ -0,0 +1,242 @@
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package triggerwatch
+
+import (
+	"fmt"
+	"syscall"
+	"time"
+	"unsafe"
+
+	// TODO:UC20: not packaged, reimplement the minimal things we need?
+	evdev "github.com/gvalkov/golang-evdev"
+
+	"github.com/snapcore/snapd/logger"
+)
+
+type keyEvent struct {
+	Dev triggerDevice
+	Err error
+}
+
+type triggerEventFilter struct {
+	Key string
+}
+
+var (
+	strToKey = map[string]int{
+		"KEY_ESC": evdev.KEY_ESC,
+		"KEY_1":   evdev.KEY_1,
+		"KEY_2":   evdev.KEY_2,
+		"KEY_3":   evdev.KEY_3,
+		"KEY_4":   evdev.KEY_4,
+		"KEY_5":   evdev.KEY_5,
+		"KEY_6":   evdev.KEY_6,
+		"KEY_7":   evdev.KEY_7,
+		"KEY_8":   evdev.KEY_8,
+		"KEY_9":   evdev.KEY_9,
+		"KEY_0":   evdev.KEY_0,
+	}
+	evKeyCapability = evdev.CapabilityType{Type: evdev.EV_KEY, Name: "EV_KEY"}
+
+	// hold time needed to trigger the event
+	holdToTrigger = 2 * time.Second
+)
+
+func init() {
+	trigger = &evdevInput{}
+}
+
+type evdevKeyboardInputDevice struct {
+	keyCode uint16
+	dev     *evdev.InputDevice
+}
+
+func (e *evdevKeyboardInputDevice) probeKeyState() (bool, error) {
+	// XXX: evdev defines EVIOCGKEY using MAX_NAME_SIZE which is larger than
+	// what is needed to store the key bitmap with KEY_MAX bits, but we need
+	// to play along since the value is already encoded
+	keyBitmap := new([evdev.MAX_NAME_SIZE]byte)
+
+	// obtain the large bitmap with all key states
+	// https://elixir.bootlin.com/linux/v5.5.5/source/drivers/input/evdev.c#L1163
+	_, _, err := syscall.RawSyscall(syscall.SYS_IOCTL, e.dev.File.Fd(), uintptr(evdev.EVIOCGKEY), uintptr(unsafe.Pointer(keyBitmap)))
+	if err != 0 {
+		return false, err
+	}
+	byteIdx := e.keyCode / 8
+	keyMask := byte(1 << (e.keyCode % 8))
+	isDown := keyBitmap[byteIdx]&keyMask != 0
+	return isDown, nil
+}
+
+func (e *evdevKeyboardInputDevice) WaitForTrigger(ch chan keyEvent) {
+	logger.Noticef("%s: starting wait", e)
+
+	// XXX: do not mess with setting the key repeat rate, as it's cumbersome
+	// and golang-evdev SetRepeatRate() parameter order is actually reversed
+	// wrt. what the kernel does. The evdev interprets EVIOCSREP arguments
+	// as (delay, repeat)
+	// https://elixir.bootlin.com/linux/latest/source/drivers/input/evdev.c#L1072
+	// but the wrapper is passing is passing (repeat, delay)
+	// https://github.com/gvalkov/golang-evdev/blob/287e62b94bcb850ab42e711bd74b2875da83af2c/device.go#L226-L230
+
+	holdState, err := e.probeKeyState()
+	if err != nil {
+		ch <- keyEvent{Err: fmt.Errorf("cannot obtain initial key state: %v", err), Dev: e}
+	}
+	if holdState {
```

Comment:
personally the rest of the code would read much simpler at first if we had this as `keyDown` or maybe even `keyDown`, because whether the key is held or not makes sense for the variable name but adding `state` to the end adds more meaning to me that gets confusing. Maybe it's just me and this makes sense to everybody else however...

Reply:
Good point, updated the name